components/raftstore/src/store/peer.rs

```diff
@@ -1609,6 +1644,86 @@ impl Peer {
+        if let Some(req) = self.batch_cmd.request.take() {
+            self.batch_cmd.batch_size = 0;
+            let mut cbs = mem::replace(&mut self.batch_cmd.callbacks, vec![]);
+            let term = req.get_header().get_term();
+            if let Err(e) = util::check_term(&req, self.term()) {
+                self.batch_callback(cbs, term, e);
+                return;
+            }
+            match self.propose_normal(ctx, req) {
+                Err(e) => {
+                    self.batch_callback(cbs, term, e);
+                }
+                Ok(idx) => {
+                    let meta = ProposalMeta {
+                        index: idx,
+                        term: self.term(),
+                        renew_lease_time: None,
+                    };
+                    if cbs.len() > 1 {
+                        ctx.raft_metrics.propose.batch += cbs.len() - 1;
+                        self.post_propose(
+                            ctx,
+                            meta,
+                            false,
+                            Callback::Write(Box::new(move |resp| {
+                                let mut last_index = 0;
+                                let has_error = resp.response.get_header().has_error();
+                                for (cb, req_num) in cbs {
+                                    let next_index = last_index + req_num;
+                                    let mut cmd_resp = RaftCmdResponse::default();
+                                    cmd_resp.set_header(resp.response.get_header().clone());
+                                    if !has_error {
+                                        cmd_resp.set_responses(
+                                            resp.response.get_responses()[last_index..next_index]
+                                                .into(),
+                                        );
+                                    }
+                                    cb.invoke_with_response(cmd_resp);
+                                    last_index = next_index;
+                                }
+                            })),
+                        );
+                    } else {
+                        let (cb, _) = cbs.pop().unwrap();
+                        self.post_propose(ctx, meta, false, cb);
+                    }
+                }
+            }
+        }
+    }
+
+    fn should_propose_batch_request<T, C>(
+        &mut self,
+        ctx: &mut PollContext<T, C>,
+        req: &RaftCmdRequest,
+    ) -> bool {
+        if let Some(batch_req) = self.batch_cmd.request.as_ref() {
+            if batch_req.get_header() != req.get_header() {
+                return true;
+            }
+            if f64::from(self.batch_cmd.batch_size) > ctx.cfg.raft_entry_max_size.0 as f64 * 0.4 {
+                return true;
+            }
+            if batch_req.get_requests().len() > MAX_BATCH_KEY_NUM {
```

Comment:
One request is not equal to one key, so `MAX_BATCH_KEY_NUM` here is not appropriate.

Reply:
Emmm? The interface `async_write` turns every `Modify` to a `Request`, why there may be more than one key?